packages/compass-aggregations/src/modules/pipeline-builder/pipeline-mode.ts

```diff
@@ -1,13 +1,31 @@
 import type { Reducer } from 'redux';
 
-type PipelineMode = 'builder-ui';
+export type PipelineMode = 'builder-ui' | 'as-text';
+
+export enum ActionTypes {
+  PipelineModeToggled = 'compass-aggregations/pipelineModeToggled',
+}
+
+export type PipelineModeToggledAction = {
+  type: ActionTypes.PipelineModeToggled;
+  mode: PipelineMode;
+};
 
 type State = PipelineMode;
+type Actions = PipelineModeToggledAction;
 
 export const INITIAL_STATE: State = 'builder-ui';
 
-const reducer: Reducer<State> = (state = INITIAL_STATE) => {
+const reducer: Reducer<State, Actions> = (state = INITIAL_STATE, action) => {
```

Comment:
Yes, the redux does pass all the actions through the reducer, but it depends on how we compose the reducer function. we can handle various `action.type`s there and accordingly type the action parameter, which will ensure handling of known `action.type`s. 

The only benefit of using `isAction` over typing `action` is when I have to handle new action type, I don't have to extend the type of `action`. 

I made the change though :)

Reply:
That's true, but you are relying on a fact that nobody will change this code to a potentially broken one after your changes. It would also be hard to spot that something is wrong at a glance. To be clear what I mean: [here's a broken reducer example](https://www.typescriptlang.org/play?#code/C4TwDgpgBAggxsAlgewHYGcoF4oHIBmyyuUAPngEYCGATrgFD2iSyojxJrZTrA2KoA5gG5G+AK6oEKVFBoQAJuLgQaACl5VgEAFxRU4gLYVVAGihVpaPRxnoAlHoPHVUAN70oPAO6JgcAAs1S05Ue3dPLyg4KnRoAiJcHUiouQhgcRpZAEZRVJi4yloklKj5DKyoACY8qIB6OqgAURoaZBpzVGQoBQh8KnEAG2Bzb2hyzNRFKEle-AFpzW0oATlJJEMIUoaoExjxQpCZFcwu4ChBxEM-aeBuhOIyIrpIgF96d-oJKVC0pRUaFUNMAtLp9EYTB0LFZUDY2LY0I5wS4aBEvOhfP4gkdEWiogV4oRiMlUl4JpVcqUCc8SqS0hVZDVto0AMpUTZQVRtKEUcTnAAqLJ4cHk7MwWhm4ooyD5K2Abw+QA) that TS thinks is totally valid due to actions being too narrow